### ISSUE TITLE
[TLS] bytes(TLS_Ext_CSR) gives wrong result when TLS_Ext_CSR.len=0

### ISSUE THREAD

**Issue by micpez** [secdev/scapy#1721]
#### Brief description
bytes(TLS_Ext_CSR) returns one additional byte when with TLS_Ext_CSR.len=0

The reason should be that:
- raw_packet_cache is not considered in `_GenericTLSSessionInheritance.__bytes__()`
- `<Field (TLS_Ext_CSR).stype>.i2m()` method returns 0 even if val is None.

#### Environment
-  Scapy version: 2.4.0.dev601
-  Python: 3.6.5
-  Operating System: Ubuntu

#### How to reproduce
```
ext_csr = TLS_Ext_CSR(b'\x00\x05\x00\x00')
>>> ext_csr.show()
###[ TLS Extension - Certificate Status Request ]### 
  type      = status_request
  len       = 0
  stype     = None
  \req       \
```

#### Actual result
```
>>> bytes(ext_csr)
b'\x00\x05\x00\x00\x00'
```

#### Expected result
```
>>> ext_csr.raw_packet_cache
b'\x00\x05\x00\x00'
```

As a consequence, the length (and bytes) of the TCP payload is also wrong, e.g. in the example in the picture, instead of length=1360:
```
>>> len(pkt[TCP].payload)
1361
```
![image](https://user-images.githubusercontent.com/45465437/49289574-79cb9880-f4a4-11e8-985d-3354c4920c6d.png)

#### Related resources

> traces / sample pcaps (stripped to the relevant frames), related standards, RFCs or other resources

**Comment by gpotter2** [comment]
Hi, thanks for reporting.

The `raw_packet_cache` is already fixed in https://github.com/secdev/scapy/pull/1705
I’ll add the second fix to it too. We’re updating the TLS 1.3 implementation anyways

**Comment by guedou** [comment]
I have been reading the RFC and I think that current Scapy behavior is correct. I believe that the best solution is to use `TLS_Ext_Unknown(type=5, len=0)` to generate such extension.

Is is something that you are trying to inject, of sniff from the network? Could you share some packets? I did not manage to capture similar ones.

**Comment by micpez** [comment]
Sorry, I cannot retrieve the packet. I was sniffing from the network, the picture is a real case.

**Comment by guedou** [comment]
If I understand correctly, you sniffed the Server Hello message,
dissected it with Scapy, then tried to build it again?

The real extension was 4 bytes, but Scapy returned 5. Am I correct?

**Comment by guedou** [comment]
Thanks for the report.

I will try to find a correct fix. As of today, Scapy encodes the Certificate Status Request on 5 bytes. It still did not find an explanation in RFCs but other implementations seem to consider that empty extensions exist and are encoder on 4 bytes.

**Comment by micpez** [comment]
Scapy encodes on 5 bytes due to:
* raw_packet_cache is not considered in _GenericTLSSessionInheritance.__bytes__()
* <Field (TLS_Ext_CSR).stype>.i2m() method returns 0 even if val is None.

I do not know if this is wanted due to an RFC or not.

**Comment by Asmeble** [comment]
this needs to be fixed immediately.

**Comment by gpotter2** [comment]
Fixed (indirectly) by #4012.
